src/Middleware/RateLimiting/src/RateLimiterOptionsExtensions.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading.RateLimiting;
+
+namespace Microsoft.AspNetCore.RateLimiting;
+
+/// <summary>
+/// Extension methods for the RateLimiting middleware options.
+/// </summary>
+public static class RateLimiterOptionsExtensions
+{
+    /// <summary>
+    /// Adds a new <see cref="TokenBucketRateLimiter"/> with the given <see cref="TokenBucketRateLimiterOptions"/> to the <see cref="RateLimiterOptions"/>.
+    /// </summary>
+    /// <param name="options">The <see cref="RateLimiterOptions"/> to add a limiter to.</param>
+    /// <param name="policyName">The name that will be associated with the limiter.</param>
+    /// <param name="tokenBucketRateLimiterOptions">The <see cref="TokenBucketRateLimiterOptions"/> to be used for the limiter.</param>
+    /// <returns>This <see cref="RateLimiterOptions"/>.</returns>
+    public static RateLimiterOptions AddTokenBucketLimiter(this RateLimiterOptions options, string policyName, TokenBucketRateLimiterOptions tokenBucketRateLimiterOptions)
+    {
+        return options.InternalAddPolicy(policyName, context =>
+        {
+            return RateLimitPartition.CreateTokenBucketLimiter((DefaultKeyType)new DefaultKeyType<PolicyNameKey>(new PolicyNameKey() { PolicyName = policyName }),
```

Comment:
This is a per request call, we should avoid these allocations every request. The DefaultKeyType should be created once at startup and reused.